### PR TITLE
R implementation of sqlParseVariablesImpl

### LINKED ISSUE #82: DBI depending on Rcpp?
I would like to suggest DBI not depending on Rcpp. The rationale here is that an interface package should be as independent as possible, and Rcpp is a huge dependency. Also, when considering alternative R implementations, having DBI depend on it is not desirable since these alternative implementations are unlikely to support Rcpp.

On closer review, I see that Rcpp is only used to parse SQL strings. IMHO, this is also not really a job for DBI to semantically analyze the query strings.

### PR DESCRIPTION
**Pull request by hannesmuehleisen**
Fixes #82.

### CODE REVIEW

**Comment by hannesmuehleisen** [comment]
@hadley any comments?

File: R/interpolate.R [hunk 1]
```diff
@@ -118,3 +117,76 @@ sqlCommentSpec <- function(start, end, endRequired) {
...
+    if (in_comment == 0L) {
+      # check all quote defintions, they can only be single characters
+      for(q in seq_along(quotes)) {
+        if (in_quote == 0L) {
+          quote_start_char <- quotes[[q]][[1]]
+          if (substr(sql_string, i, i) == quote_start_char)  {
+            in_quote <- q
+            break
+          }
+      } else {
```
Comment (krlmlr) [inline R/interpolate.R]: Indentation looks odd here.

File: R/interpolate.R [hunk 2]
```diff
@@ -118,3 +117,76 @@ sqlCommentSpec <- function(start, end, endRequired) {
...
+      for(c in seq_along(comments)) {
+        if (in_comment == 0L) {
+          comment_start_string <- comments[[c]][[1]]
+          comment_start_length <- nchar(comment_start_string) - 1
+          if (substr(sql_string, i, i + comment_start_length) == comment_start_string)  {
+            in_comment <- c
+            i <- i + comment_start_length
+            break
+          }
+      } else {
```
Comment (krlmlr) [inline R/interpolate.R]: Same here.

File: NAMESPACE [hunk 3]
```diff
@@ -71,6 +71,3 @@ exportMethods(sqlCreateTable)
 exportMethods(sqlData)
 exportMethods(sqlInterpolate)
 exportMethods(sqlParseVariables)
-import(methods)
```
Comment (krlmlr) [inline NAMESPACE]: We might still need methods, no?

**Comment by krlmlr** [comment]
What's the run time of the new code on a 1000, 10000, 100000-char string, compared to the old code?

**Comment by hannesmuehleisen** [comment]
did a roxygen run, will check indents next

**Comment by hannesmuehleisen** [comment]
whee the checks have passed. quick benchmark next.

**Comment by hannesmuehleisen** [comment]
Tested with the following line on my MacBook:

    system.time(sqlParseVariables(ANSI(), paste0(rep("S", 100000), collapse="")))

Runtimes old code:
1000-char: 0.00s
10000-char: 0.00s
100000-char: 0.02

Runtimes new code:
1000-char: 0.03s
10000-char: 0.48s
100000-char: 34.06s (!)

So, as to expected, its way slower. But I would say since this is just a fallback and queries are expected to be shorter than 10K characters its not going to matter much.

**Comment by krlmlr** [comment]
It might turn out that this function will be called for all parametrized queries, regardless of backend support, to provide a backend-independent syntax for parametrized queries. Anyway, the C++ implementation might as well live in a separate package.

The apparent non-linear behavior of the R implementation bothers me. Any chance we can get rid of it, perhaps through some clever strsplit() preprocessing?

Could you please post/upload the testing code?

**Comment by hannesmuehleisen** [comment]
Presence of quotes/comments only changes a state bits, so I would not expect that to make any difference.

**Comment by hannesmuehleisen** [comment]
Strange, when run in Rstudio it only takes 4.23 seconds on the largest string as one would expect form linearity.

**Comment by hannesmuehleisen** [comment]
looks like `substr` is the main culprit here. Will try to optimize.

**Comment by hannesmuehleisen** [comment]
Got the slowest time (100K char string) down to 1.5 sec. Pushing today...

**Comment by krlmlr** [comment]
This is very nice indeed! Looking forward to it.

**Comment by hannesmuehleisen** [comment]
Runtimes faster new code:
1000-char: 0.01s
10000-char: 0.14s
100000-char: 1.52s 
1000000-char: 14.39s

File: R/interpolate.R [hunk 4]
```diff
@@ -118,3 +117,86 @@ sqlCommentSpec <- function(start, end, endRequired) {
 sqlQuoteSpec <- function(start, end, escape = "", doubleEscape = TRUE) {
   list(start, end, escape, doubleEscape)
 }
+
+#' @export
+#' @rdname sqlParseVariables
+sqlParseVariablesImpl <- function(sql, quotes, comments) {
+  sql_arr <- strsplit(as.character(sql), "", fixed=T)[[1]]
```
Comment (krlmlr) [inline R/interpolate.R]: Could you please use `TRUE` instead of `T`?

File: R/interpolate.R [hunk 5]
```diff
@@ -118,3 +117,86 @@ sqlCommentSpec <- function(start, end, endRequired) {
...
+  var_pos_start <- integer()
+  var_pos_end <- integer()
+
+  # prepare comments
+  for(c in seq_along(comments)) {
+    comments[[c]][[1]] <- strsplit(comments[[c]][[1]], "", fixed=T)[[1]]
+    comments[[c]][[2]] <- strsplit(comments[[c]][[2]], "", fixed=T)[[1]]
+  }
+  
+  var_chars <- c(LETTERS, tolower(LETTERS), 0:9, "_")
```
Comment (krlmlr) [inline R/interpolate.R]: tolower(LETTERS) == letters

File: R/interpolate.R [hunk 6]
```diff
@@ -118,3 +117,86 @@ sqlCommentSpec <- function(start, end, endRequired) {
...
+      for(q in seq_along(quotes)) {
+        if (in_quote == 0L) {
+          quote_start_char <- quotes[[q]][[1]]
+          if (identical(sql_arr[[i]], quote_start_char))  {
+            in_quote <- q
+            break
+          }
+        } else {
+          # only check the end of the active quote definition
+          # TODO: support end quote escaping (e.g. \")
```
Comment (krlmlr) [inline R/interpolate.R]: Was end quote escaping supported by the Rcpp code?

**Comment by hannesmuehleisen** [comment]
Not sure about the end quote escaping.

File: man/sqlParseVariables.Rd [hunk 7]
```diff
@@ -18,16 +16,10 @@ sqlParseVariables(con, sql, ...)
 sqlCommentSpec(start, end, endRequired)
 
 sqlQuoteSpec(start, end, escape = "", doubleEscape = TRUE)
+
+sqlParseVariablesImpl(sql, quotes, comments)
 }
 \arguments{
-\item{sql}{SQL to parse (a character vector of length 1)}
```
Comment (krlmlr) [inline man/sqlParseVariables.Rd]: Please re-add parameter documentation to the R source.

**Comment by hannesmuehleisen** [comment]
added doc

File: R/interpolate.R (hunk 6 above)
Comment (krlmlr) [inline R/interpolate.R]: I think there was support for it, although not tested: https://codecov.io/github/rstats-db/DBI/src/sqldelim.cpp?ref=ef1a72121ed1f34e5bf94a913ea3410d9b289f42#l-61 . I'd like to review the old code and have it fully covered by tests before proceeding here. Also, to me it looks like the construct `"SELECT ? FROM A"` will be interpolated in a strange way by the C++ implementation.

**Comment by krlmlr** [comment]
Thanks for the PR. Just two more nits:

- Could you please remove spaces before EOL?
- Please add spaces before and after `=`.

Would you like to look into improving test coverage for the existing code? Perhaps the new tests from this PR already do.

**Comment by hannesmuehleisen** [comment]
I had added two test cases that I found to be tricky. Will look into test coverage.

**Comment by hadley** [comment]
This looks like a reasonable approach to me, although I didn't read it closely.

**Comment by hannesmuehleisen** [comment]
Added some corner test cases and checked test coverage, sqlParseVariablesImpl() is completely covered.

**Comment by krlmlr** [comment]
@hannesmuehleisen: I've added full test coverage for the C++ implementation to master. Could you please merge/rebase and make sure that tests pass?

**Comment by hannesmuehleisen** [comment]
will do

**Comment by hannesmuehleisen** [comment]
Ok should be all well by now, test cases pass. Runtime is unchanged.

File: man/sqlParseVariables.Rd [hunk 8]
```diff
@@ -1,7 +1,7 @@
 % Generated by roxygen2: do not edit by hand
-% Please edit documentation in R/RcppExports.R, R/interpolate.R
+% Please edit documentation in R/interpolate.R
 \docType{methods}
-\name{sqlParseVariablesImpl}
```
Comment (krlmlr) [inline man/sqlParseVariables.Rd]: Why is the alias gone?

File: R/interpolate.R [hunk 9]
```diff
@@ -29,18 +29,17 @@ setGeneric("sqlInterpolate", function(`_con`, `_sql`, ...) {
 setMethod("sqlInterpolate", "DBIConnection", function(`_con`, `_sql`, ...) {
   sql <- `_sql`
   pos <- sqlParseVariables(`_con`, sql)
-
+
```
Comment (krlmlr) [inline R/interpolate.R]: Nitpick... Current RStudio trims whitespace, I've updated the project settings in d2c7a78f90d21385860adba04b4bd72f19e566fb.

**Comment by krlmlr** [comment]
Run time is still quadratic in the number of variables, but I guess we can live with that:

```
> s <- paste0(rep("?a", 10000), collapse=""); system.time(sqlParseVariables(ANSI(), s))
   user  system elapsed 
  0.520   0.000   0.521 
> s <- paste0(rep("?a", 20000), collapse=""); system.time(sqlParseVariables(ANSI(), s))
   user  system elapsed 
  1.840   0.000   1.846 
```

**Comment by hannesmuehleisen** [comment]
Not using Rstudio here.

**Comment by hannesmuehleisen** [comment]
No idea what happened to the alias. This file is generated after all.

**Comment by krlmlr** [comment]
It's not the alias, just the name (which is good as it is now).

Looks good to me. @hadley: Any further comments?

File: R/interpolate.R [hunk 10]
```diff
@@ -118,3 +117,105 @@ sqlCommentSpec <- function(start, end, endRequired) {
...
+  for(q in seq_along(quotes)) {
+   quotes[[q]][[5]] <- nchar(quotes[[q]][[3]]) > 0L
+  }
+
+  var_chars <- c(LETTERS, letters, 0:9, "_")
+  in_quote <- 0L
+  in_comment <- 0L
+  i <- 1
+  while(i <= length(sql_arr)) {
+    # only check for variables if neither commented nor quoted
```
Comment (hadley) [inline R/interpolate.R]: Indenting looks slightly off here

**Comment by hannesmuehleisen** [comment]
added a missing space.

**Comment by hadley** [comment]
My main concern the C++ used a state machine which is easy to reason about, and easy to adapt to changing requirements. I can't easily identify the strategy used in the new code, which means that if requirements change, @hannesmuehleisen will have to make the changes.   That said, I don't think this code is likely to change much in the future, so it's not a huge concern.

**Comment by hannesmuehleisen** [comment]
Can rewrite to state machine if it helps

**Comment by hadley** [comment]
If you're willing, I'd really appreciate it.

**Comment by hannesmuehleisen** [comment]
Sure, be back in an hour or so.

**Comment by hannesmuehleisen** [comment]
Ok, its now a state machine. Runtime is unchanged.

**Comment by hadley** [comment]
LGTM

File: R/interpolate.R [hunk 11]
```diff
@@ -118,3 +117,121 @@ sqlCommentSpec <- function(start, end, endRequired) {
...
+
+  # internal helper variables
+  quote_spec_offset <- 0L
+  comment_spec_offset <- 0L
+  sql_variable_start <- 0L
+  sql_variable_end <- 0L
+
+  # prepare comments & quotes for quicker comparisions
+  for(c in seq_along(comments)) {
+    comments[[c]][[1]] <- strsplit(comments[[c]][[1]], "", fixed = TRUE)[[1]]
```
Comment (krlmlr) [inline R/interpolate.R]: The [[n]] access is not very intuitive, but fastest:

```
a <- list(start = "a", end = "b", endRequired = FALSE); microbenchmark::microbenchmark(a[[3]], a[["start"]], a[["end"]], a$s, a$start, a$e, a$end, times = 100000)
Unit: nanoseconds
         expr min  lq     mean median  uq     max neval cld
       a[[3]] 125 142 201.7776    153 167   14234 1e+05  a 
 a[["start"]] 136 156 215.8544    167 184   21100 1e+05  a 
   a[["end"]] 145 164 225.6624    173 191   62662 1e+05  a 
          a$s 186 206 299.5662    218 297   43130 1e+05  ab
      a$start 151 171 265.3310    183 261   22357 1e+05  ab
          a$e 188 207 371.8202    219 298 6883402 1e+05   b
        a$end 167 188 281.5594    201 278   21501 1e+05  ab
```

I wonder how much this affects overall performance (we're talking nanoseconds here). Let's look at this in a separate PR.

File: R/interpolate.R [hunk 12]
```diff
@@ -118,3 +117,121 @@ sqlCommentSpec <- function(start, end, endRequired) {
...
+    ) # </switch>
+  } # </while>
+
+  if (quote_spec_offset > 0L) {
+    stop("Unterminated literal")
+  }
+  if (comment_spec_offset > 0L && comments[[comment_spec_offset]][[3]]) {
+    stop("Unterminated comment")
+  }
+  list(start = as.integer(var_pos_start), end = as.integer(var_pos_end))
```
Comment (krlmlr) [inline R/interpolate.R]: If you're careful to keep `i` an integer, the coercion here is probably unnecessary. Again, separate PR.

**Comment by krlmlr** [comment]
Thanks for your efforts.

**Comment by hannesmuehleisen** [comment]
Whee, thanks for merging!

**Comment by jangorecki** [comment]
@hannesmuehleisen thanks for this PR! just spotted that I can upgrade DBI and was worried as I was sure DBI will depend on Rcpp based on #40. Great surprise to see it is still lightweight.

**Pull request merged**